src/dawn/Optimizer/PassSetBoundaryCondition.cpp

```diff
@@ -0,0 +1,337 @@
+    StencilIDsVisited_.push_back(
+        stencilInstantiation->getStencilCallToStencilIDMap().find(stencilcall)->second);
+  }
+
+  auto calculateHaloExtents = [&](std::string fieldname) {
+
+    Extents fullExtent;
+    // Did we already apply a BoundaryCondition for this field?
+    // This is the first time we apply a BC to this field, we traverse all stencils that were
+    // applied before
+    std::vector<int> stencilIDsToVisit(StencilIDsVisited_);
+    if(StencilBCsApplied_.count(fieldname) == 0) {
+    } else {
+      for(int traveresedID : StencilBCsApplied_.find(fieldname)->second) {
+        stencilIDsToVisit.erase(
+            std::remove(stencilIDsToVisit.begin(), stencilIDsToVisit.end(), traveresedID),
+            stencilIDsToVisit.end());
+      }
+    }
+    for(auto& stencil : stencilInstantiation->getStencils()) {
+      for(const int& ID : stencilIDsToVisit) {
+        if(ID == stencil->getStencilID()) {
+          fullExtent.merge(
+              analyzeStencilExtents(stencil, stencilInstantiation->getAccessIDFromName(fieldname)));
+          if(StencilBCsApplied_.count(fieldname) == 0) {
+            StencilBCsApplied_.emplace(fieldname, std::vector<int>{stencil->getStencilID()});
+          } else {
+            StencilBCsApplied_.find(fieldname)->second.push_back(stencil->getStencilID());
+          }
+          break;
+        }
+      }
+    }
+    return fullExtent;
+  };
+
+  // Loop through all the StmtAccessPair in the stencil forward
+  for(const auto& stencilPtr : stencilInstantiation->getStencils()) {
+    Stencil& stencil = *stencilPtr;
+    DAWN_LOG(INFO) << "analyzing stencil " << stencilInstantiation->getName();
+    std::unordered_map<int, Extents> stencilDirtyFields;
+    stencilDirtyFields.clear();
+
+    for(const auto& multiStagePtr : stencil.getMultiStages()) {
+      MultiStage& multiStage = *multiStagePtr;
+      for(auto stageIt = multiStage.getStages().begin(); stageIt != multiStage.getStages().end();
+          ++stageIt) {
+        Stage& stage = (**stageIt);
+        for(const auto& domethod : stage.getDoMethods()) {
+          for(const auto& stmtAccess : domethod->getStatementAccessesPairs()) {
+            Accesses& acesses = *(stmtAccess->getAccesses());
+            const auto& allReadAccesses = acesses.getReadAccesses();
+            const auto& allWriteAccesses = acesses.getWriteAccesses();
+
+            // ReadAccesses can trigger Halo-Updates and Boundary conditions if the following
+            // criteria are fullfilled:
+            // It is a Field (ID!=-1) and we had a write before from another stencil (is in
+            // dirtyFields)
+            for(const auto& readacccess : allReadAccesses) {
+              int originalID = getOriginalID(readacccess.first);
+              if(originalID == -1)
+                continue;
+              auto idWithExtents = dirtyFields.find(originalID);
+              if(idWithExtents != dirtyFields.end()) {
```

Comment:
I would suggest to use more the early exit pattern with return here. In general blocks statements which are nested and large like here are hard to read
```
int readAccessID = readacccess.first;
Extent readAccessExtent = readacccess.second;
if(! dirtyFields.count(originalID)) continue;
if( readAcccessExtent.isHorizontalPointwise() ||
                   stencilInstantiation->getCachedVariableSet().count(readAcccessID) )
continue;
```
Also above I tried to avoid `first` and `second` for readability. Best would be to have a proper structure in `readAccesses_ ` of 
 `Accesses`, that is also probably separate work.